File: frontend/upload-lib/ts/sanitize.test.ts
import { describe, expect, it } from "vitest";
import { sanitizeFilename } from "./sanitize";

describe("sanitizeFilename", () => {
  it("normal", () => {
    expect(sanitizeFilename("photo.jpg")).toBe("photo.jpg");
  });
  it("path traversal unix", () => {
    expect(sanitizeFilename("../../etc/passwd")).toBe("passwd");
  });
  it("path traversal nested", () => {
    expect(sanitizeFilename("foo/../../bar.txt")).toBe("bar.txt");
  });
  it("dangerous chars", () => {
    expect(sanitizeFilename('a:b*c?d|e<f>g"h.txt')).toBe("abcdefgh.txt");
  });
  it("null byte", () => {
    expect(sanitizeFilename("file\x00name.txt")).toBe("filename.txt");
  });
  it("dots only after sanitize", () => {
    expect(sanitizeFilename("..")).toBe("unnamed_file");
  });
  it("empty", () => {
    expect(sanitizeFilename("")).toBe("unnamed_file");
  });
  it("dot", () => {
    expect(sanitizeFilename(".")).toBe("unnamed_file");
  });
  it("backslash path", () => {
    expect(sanitizeFilename("..\\..\\secret.txt")).toBe("secret.txt");
  });
});

File: backend/pkg/upload/validation/sanitize_test.go
package validation

import "testing"

func TestSanitizeFilename(t *testing.T) {
	tests := []struct {
		name  string
		input string
		want  string
	}{
		{name: "normal", input: "photo.jpg", want: "photo.jpg"},
		{name: "path traversal unix", input: "../../etc/passwd", want: "passwd"},
		{name: "path traversal nested", input: "foo/../../bar.txt", want: "bar.txt"},
		{name: "dangerous chars", input: `a:b*c?d|e<f>g"h.txt`, want: "abcdefgh.txt"},
		{name: "null byte", input: "file\x00name.txt", want: "filename.txt"},
		{name: "dots only after sanitize", input: "..", want: "unnamed_file"},
		{name: "empty", input: "", want: "unnamed_file"},
		{name: "dot", input: ".", want: "unnamed_file"},
		{name: "backslash path", input: `..\..\secret.txt`, want: "secret.txt"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := SanitizeFilename(tt.input)
			if got != tt.want {
				t.Fatalf("SanitizeFilename(%q) = %q, want %q", tt.input, got, tt.want)
			}
		})
	}
}
